analytics/analytics.go

```diff
@@ -0,0 +1,258 @@
+		return false
+	}
+	defer consentFile.Close()
+
+	contents, err := ioutil.ReadAll(consentFile)
+	if err != nil {
+		return false
+	}
+
+	var consent cliTrackingConsent
+	err = json.Unmarshal(contents, &consent)
+
+	return err == nil
+}
+
+// ReadConsent finds if consent has been given, either from file or by prompt.
+func ReadConsent() {
+	if !IsConsentRecorded() {
+		PromptForConsent()
+	}
+
+	if _, err := os.Stat(consentPath); err != nil {
+		return
+	}
+
+	consentFile, err := os.Open(consentPath)
+	if err != nil {
+		return
+	}
+	defer consentFile.Close()
+
+	var consent cliTrackingConsent
+	contents, err := ioutil.ReadAll(consentFile)
+	if err != nil {
+		return
+	}
+	err = json.Unmarshal(contents, &consent)
+	if err != nil {
+		return
+	}
+
+	ConsentGiven = consent.ConsentGiven
+}
+
+// Println formats using the default formats for its operands and writes to output.
+// Output can be overridden for testing purposes by setting: analytics.out
+// It returns the number of bytes written and any write error encountered.
+func Println(a ...interface{}) (n int, err error) {
+	return fmt.Fprintln(out, a...)
+}
+
+// Printf formats according to a format specifier and writes to standard output.
+// Output can be overridden for testing purposes by setting: analytics.out
+// It returns the number of bytes written and any write error encountered.
+func Printf(format string, a ...interface{}) (n int, err error) {
+	return fmt.Fprintf(out, format, a...)
+}
+
+// PromptForConsent interactively prompts the user for consent
+func PromptForConsent() {
+	Printf("👋 Hi!\n\n")
+	Printf("Thanks for using the Section CLI.\n\n")
+	Printf("We're still working out how to create the best experience for you.\n\n")
+	Printf("We'd like to collect some anonymous info about how you use the CLI.\n\n")
+	Printf("Are you OK with this? [y/N] ")
+
+	reader := bufio.NewReader(in)
+	text, err := reader.ReadString('\n')
+	if err != nil {
+		Println("Error: unable to read your response. Exiting.")
+		os.Exit(2)
+	}
+	text = strings.Replace(text, "\n", "", -1) // convert CRLF to LF
+
+	if strings.EqualFold(text, "y") {
+		Printf("\nThank you!\n")
+		ConsentGiven = true
+	} else {
+		ConsentGiven = false
+		Printf("\nNo worries! We won't ask again.\n")
+	}
+
+	WriteConsent()
+}
+
+// WriteConsent writes the current consent state to a persistent file
+func WriteConsent() {
+	c := cliTrackingConsent{ConsentGiven: ConsentGiven}
+	json, err := json.Marshal(c)
+	if err != nil {
+		Println("Error: unable to record consent. Exiting.")
+		os.Exit(2)
+	}
+	err = ioutil.WriteFile(consentPath, json, 0644)
+	if err != nil {
+		Println("Error: unable to record consent. Exiting.")
+		os.Exit(2)
+	}
+}
+
+// Submit submits an analytics event to Section
+//
+// Behavior is determined by consent:
+//
+// if consent not given {
+// 	prompt for consent
+// }
+// if consent == true {
+// 	submit analytics
+// }
+func Submit(e Event) (err error) {
+	ReadConsent()
+	if !ConsentGiven {
+		return err
```

Comment:
Isn't error `nil` in this case? You don't assign anything to it. Am I missing something?

Reply:
Yep, this is totally fine behavior.

If there is no consent given, `Submit` should return immediately with a `nil` error.